src/display.cpp

```diff
@@ -2739,20 +2739,28 @@ void display::draw_hex(const map_location& loc)
 						point isize = image::get_size(ov.image, image::HEXED);
 						std::string ipf = ov.image;
 
-						if(ov.submerge) {
-							// Adjust submerge appropriately
-							double sub = submerge * ov.submerge;
-							// Shift the image so the waterline remains static.
-							// This is so that units swimming there look okay.
-							int shift = isize.y * (sub - submerge);
-							add_submerge_ipf_mod(ipf, isize.y, sub, shift);
-						}
-
-						const texture tex = ov.image.find("~NO_TOD_SHIFT()") == std::string::npos
+						texture tex = ov.image.find("~NO_TOD_SHIFT()") == std::string::npos
 							? image::get_lighted_texture(ipf, lt)
 							: image::get_texture(ipf, image::HEXED);
 
-						drawing_buffer_add(LAYER_TERRAIN_BG, loc, [tex](const rect& dest) { draw::blit(tex, dest); });
+						drawing_buffer_add(LAYER_TERRAIN_BG, loc, [=](const rect& dest) mutable {
+							// Adjust submerge appropriately
+							const t_translation::terrain_code terrain = get_map().get_terrain(loc);
+							const terrain_type& terrain_info = get_map().get_terrain_info(terrain);
+							const double submerge = terrain_info.unit_submerge();
+
+							submerge_data data = get_submerge_data(dest, submerge, isize, ALPHA_OPAQUE, false, false);
```

Comment:
Thinking this should go inside the if block, to avoid always calculating it for every non-submerged draw. Can handle the "dry" draw twice in each branch.

Reply:
The function immediately returns if there's nothing to do, so I can't imagine it's taking any meaningful amount of time.

Reply:
I have no idea of the context, but in _very_ performance-sensitive code, I think it's possible for _just_ a function call to make a difference (which is why inlining is a thing). But of course that assumes the compiler isn't going to inline this function call, and I don't know if that's a reasonable thing to assume.

Reply:
It already had to call multiple other functions to calculate the overlay's submerge value, so one more function call isn't going to make a difference here.

Reply:
I don't think that necessarily follows, but I also think the optimizer would probably take care of it anyway, so… it probably makes no difference.